cmd/boot-script-service/routers.go

```diff
@@ -36,37 +36,86 @@
 package main
 
 import (
+	"context"
 	"fmt"
-	base "github.com/Cray-HPE/hms-base"
 	"net/http"
+
+	base "github.com/Cray-HPE/hms-base"
+	"github.com/go-chi/chi/v5"
+	"github.com/go-chi/jwtauth/v5"
+	"github.com/lestrrat-go/jwx/jwa"
+	"github.com/lestrrat-go/jwx/jwk"
 )
 
 const (
 	baseEndpoint     = "/boot/v1"
 	notifierEndpoint = baseEndpoint + "/scn"
 	// We don't use the baseEndpoint here because cloud-init doesn't like them
-	metaDataRoute   = "/meta-data"
-	userDataRoute   = "/user-data"
-	phoneHomeRoute  = "/phone-home"
+	metaDataRoute  = "/meta-data"
+	userDataRoute  = "/user-data"
+	phoneHomeRoute = "/phone-home"
 )
 
-func initHandlers() {
-	http.HandleFunc(baseEndpoint+"/", Index)
-	// config
-	http.HandleFunc(baseEndpoint+"/bootparameters", bootParameters)
+var (
+	tokenAuth *jwtauth.JWTAuth
+)
+
+func loadPublicKeyFromURL(url string) error {
+	ctx, cancel := context.WithCancel(context.Background())
+	defer cancel()
+	set, err := jwk.Fetch(ctx, url)
+	if err != nil {
+		return fmt.Errorf("%v", err)
+	}
+	for it := set.Iterate(context.Background()); it.Next(context.Background()); {
+		pair := it.Pair()
+		key := pair.Value.(jwk.Key)
+
+		var rawkey interface{}
+		if err := key.Raw(&rawkey); err != nil {
+			continue
+		}
+
+		tokenAuth = jwtauth.New(jwa.RS256.String(), nil, rawkey)
+		return nil
+	}
+
+	return fmt.Errorf("failed to load public key: %v", err)
+}
+
+func initHandlers() *chi.Mux {
+	router := chi.NewRouter()
+	if requireAuth {
+		router.Group(func(r chi.Router) {
+			r.Use(
+				jwtauth.Verifier(tokenAuth),
+				jwtauth.Authenticator(tokenAuth),
+			)
+
+			// register config as protected route
+			r.HandleFunc(baseEndpoint+"/", Index)
+			r.HandleFunc(baseEndpoint+"/bootparameters", bootParameters)
+		})
+	} else {
+		// register config as protected route
+		router.HandleFunc(baseEndpoint+"/", Index)
+		router.HandleFunc(baseEndpoint+"/bootparameters", bootParameters)
```

Comment:
I see two `register config as protected route` comments. I assume the block in the `else` block registers the endpoints as unprotected if `requireAuth` is false?

Reply:
Looks good.